python/llm/src/bigdl/llm/transformers/qlora.py

```diff
@@ -0,0 +1,191 @@
+            out_features,
+            qtype=kwargs.get("qtype"),
+            bias=kwargs.get("bias", True),
+            conver_to_half=False,
+        )
+        LoraLayer.__init__(self, in_features=in_features, out_features=out_features)
+
+        # Freezing the pre-trained weight matrix
+        self.weight.requires_grad = False
+
+        init_lora_weights = kwargs.pop("init_lora_weights", True)
+        self.update_layer(adapter_name, r, lora_alpha, lora_dropout, init_lora_weights)
+        self.active_adapter = adapter_name
+
+    def forward(self, x: torch.Tensor):
+        result = super().forward(x)
+
+        if self.disable_adapters or self.active_adapter not in self.lora_A.keys():
+            return result
+        elif self.r[self.active_adapter] > 0:
+            result = result.clone()
+            if not torch.is_autocast_enabled():
+                expected_dtype = result.dtype
+                x = x.to(self.lora_A[self.active_adapter].weight.dtype)
+                output = (
+                    self.lora_B[self.active_adapter](
+                        self.lora_A[self.active_adapter](self.lora_dropout[self.active_adapter](x))
+                    ).to(expected_dtype)
+                    * self.scaling[self.active_adapter]
+                )
+            else:
+                output = (
+                    self.lora_B[self.active_adapter](
+                        self.lora_A[self.active_adapter](self.lora_dropout[self.active_adapter](x))
+                    )
+                    * self.scaling[self.active_adapter]
+                )
+            result += output
+        return result
+
+
+@staticmethod
+def _create_new_module(lora_config, adapter_name, target, **kwargs):
+
+    bias = kwargs.pop("bias", False)
+
+    if isinstance(target, LowBitLinear):
+        low_bit_kwargs = kwargs.copy()
+        low_bit_kwargs.update(
+            {
+                "qtype": target.qtype,
+            }
+        )
+        new_module = LoraLowBitLinear(adapter_name,
+                                      target.in_features,
+                                      target.out_features,
+                                      bias=bias,
+                                      **low_bit_kwargs)
+    else:
+        invalidInputError(False,
+                          f"Target module {target} is not supported. "
+                          f"Currently, only `LowBitLinear` are supported.")
+
+    return new_module
+
+
+from peft.tuners.lora import LoraModel
+
+
+def get_peft_model(*args, **kwargs):
+    old_create_new_module = LoraModel._create_new_module
+    LoraModel._create_new_module = _create_new_module
+    try:
+        from peft import get_peft_model as get_peft_model_original
+        model = get_peft_model_original(*args, **kwargs)
+    finally:
+        LoraModel._create_new_module = old_create_new_module
+
+    return model
+
+
+def prepare_model_for_kbit_training(model, use_gradient_checkpointing=True):
```

Comment:
do we want to set `model.is_loaded_in_4bit` to true, so as to directly reuse the original `prepare_model_for_kbit_training`

Reply:
I guess we can, but I think it might trigger some other hard-coded `bitsandbytes` related behavior in `transformers`, so this requires more testing. I can explore this option in follow up PRs.